### PR TITLE
Ability to access the current context via the service registry

### LINKED ISSUE #1937: Accessing the Context via DI
For various reasons we need to be able to get the `Context` object of the current request via DI / Service Registry.

We're using the Guice extension and our current solution is another custom extension that wraps the Guice extension and creates the request scope as in 1.x and registers the `Context` in it.

Now I came up with a much simpler (and DI framework independent) solution:

```java
// install before any route:

public class ContextAsService implements Extension {

    private static final ThreadLocal<Context> provider = new ThreadLocal<>();

    @Override
    public void install(@Nonnull Jooby application) {
        application.getServices().putIfAbsent(Context.class, (Provider<Context>) provider::get);
        application.before(provider::set);
    }
}
```

My question would be, is there any better / preferred / official way to achieve this?

### PR DESCRIPTION
**Pull request by imeszaros**
Closes #1937

I wrote a simple test case and discovered a problem. With Jetty it works fine but fails with Netty and Undertow (strange because I did my initial tests with Netty, and it was working fine...).

The problem is that Netty and Undertow dispatches the execution to a worker thread. The thread local is initialized by the event loop thread, but then the service registry is queried by the worker thread -> not found.

Do you have an idea how to resolve this?

### CODE REVIEW

**Comment by jknack** [comment]
Try with `InheritableThreadLocal` but not sure... it worked before bc you wrote a filter which run the target thread

**Comment by imeszaros** [comment]
I added a new internal api, `ExecutionCallback`. Not sure about the name though. The point is that it's called as soon as the context is passed to the final processing thread.

What do you think?

**Comment by jknack** [comment]
yea, concept is right. what if we move existing ContextInitializer after thread dispatching? Like you did with ExecutionCallback. This way we don't need both...thoughts?

**Comment by imeszaros** [comment]
Yes i was thinking about this too, but wouldn't that mess up with `HiddenMethodInitializer` and or `ContextInitializer#PROXY_PEER_ADDRESS`? They can now change which route matches with the context.

**Comment by imeszaros** [comment]
Another possibility would be to reuse `ContextInitializer` and change the methods from:

```
addInitializer(ContextInitializer)
removeInitializer(ContextInitializer)
addExecutionCallback(ExecutionCallback)
removeExecutionCallback(ExecutionCallback)
```

to:

```
addPreDispatchInitializer(ContextInitializer)
removePreDispatchInitializer(ContextInitializer)
addPostDispatchInitializer(ContextInitializer)
removePostDispatchInitializer(ContextInitializer)
```

In the `Pipiline` class i wanted to use the `decorate` method to wrap the handler, but I noticed there is one single case it is not used, with `io.jooby.internal.Pipeline#kotlinJob`. Is this intentional or is it missing?

**Comment by jknack** [comment]
Probably missing. Yea let's do that for now pre/post context initializer.

**Comment by imeszaros** [comment]
I've done the changes. I also added separate commits for three more or less related bugs I discovered.

File: modules/jooby-jetty/src/main/java/io/jooby/internal/jetty/JettyContext.java [hunk 1]
```diff
@@ -557,8 +557,8 @@ private Context sendStreamInternal(@Nonnull InputStream in) {
   @Nonnull @Override public Context onComplete(@Nonnull Route.Complete task) {
     if (listeners == null) {
       listeners = new CompletionListeners();
-      listeners.addListener(task);
     }
+    listeners.addListener(task);
```
Comment (jknack) [inline modules/jooby-jetty/src/main/java/io/jooby/internal/jetty/JettyContext.java:561]: Good catch!

**Pull request merged**